modules/packages/LinearAlgebra.chpl

```diff
@@ -605,6 +604,60 @@ private inline proc _raw(D: domain(1), i) {
 // Matrix Structure
 //
 
+
+/*
+   Return a Vector containing the diagonal elements of ``A`` if the argument ``A`` is of rank 2.
+   Return a diagonal Matrix whose diagonal contains elements of ``A`` if argument ``A`` is of rank 1.
+ */
+proc diag(A: [?Adom] ?eltType){
+  if(Adom.rank == 2) then{
+    if(!isSquare(A)) then halt("diag supports only square matrices");
+    return _diag_vec(A);  
+  }
+  else if(Adom.rank == 1) then {
+    return _diag_mat(A);
+  }
+  else{
+    compilerError("A must have rank 2 or less");
+  }
+
+}
+
+/*
+  Return a Vector containing the off diagonal elements of ``A``
+*/
+proc offdiag(A:[?Adom] ?eltType){
+  if(Adom.rank != 2) then halt("offdiag expects argument to be of rank 2");
+  if(!isSquare(A)) then halt ("offdiag supports only square matrices");
+  return _diag_vec(A,trans=true);
+}
+
+proc _diag_vec(A:[?Adom] ?eltType, trans=false){
+  var diagonal = Vector(Adom.dim(1).length);
+  if(!trans) then{
+    forall i in Adom.dim(1) do{
```

Comment:
I think there is a subtlety here that is probably best explained with an example. The input ``A`` can be any 2D array, and in Chapel, these can have indices that start and end at any point. eg. the domain of ``A`` could be ``{4..7, 4..7}``. However, when you build the vector, it will default to ``{0..3}``, following usual C numbering.... 

So in your code below, for this case, ``i`` would run from 4 to 7, while diagonal would expect 0 to 3.....

I think it's an good question of whether, in a case like this, you want the diagonal to have indices 4..7 or 0..3? I don't know.... 

Note that this applies to the cases below as well....